data_steward/cdr_cleaner/cleaning_rules/drop_participants_without_ppi_or_ehr.py

```diff
@@ -0,0 +1,92 @@
+"""
+Run the drop_participants_without_ppi_or_ehr validation clean rule.
+
+Drops all data for participants who:
+  1. have not completed "The Basics" PPI module, via the RDR
+  2. do not have any EHR data
+
+(1) is achieved by checking the observation table for children of TheBasics
+module. (2) is achieved by checking all mapping tables for all person_id tables,
+to confirm whether any data is sourced from EHR per participant.
+"""
+import logging
+
+from cdr_cleaner.cleaning_rules import drop_rows_for_missing_persons
+import common
+from constants import bq_utils as bq_consts
+from constants.cdr_cleaner import clean_cdr as clean_consts
+import resources
+
+LOGGER = logging.getLogger(__name__)
+
+BASICS_MODULE_CONCEPT_ID = 1586134
+
+SELECT_PERSON_WITH_BASICS_OR_EHR = """
+SELECT {fields}
+FROM `{project}.{dataset}.{table}` person
+LEFT JOIN (
```

Comment:
@mark-velez please also review this - my strategy is to perform an outer join across a TheBasics subquery, and every EHR table subquery to gather the criteria for the decision.

Reply:
@calbach I think the left join filtering strategy is good! In terms of syntax- we often find it difficult to understand and troubleshoot queries (SQL) interpersed with procedural logic (Python). To address this we just started to template our SQL (see [pop_retract.py](https://github.com/all-of-us/curation/blob/develop/data_steward/analytics/cdr_ops/pop_retract.py#L72)). To be sure, templates come with challenges of their own, but I'm wondering your experience, if you think using them here might improve intelligibility/testability. Either way, approach looks good. :+1:  Happy to discuss if you like.

Reply:
I like it - added. Thanks for the pointer.

Reply:
Thanks, I think this is quite a bit more intelligible!